components/camera/videosource/logging/logger.go

```diff
@@ -0,0 +1,248 @@
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/natefinch/lumberjack"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/config"
+)
+
+var filePath string
+
+func init() {
+	t := time.Now().UTC().Format(time.RFC3339)
+	t = strings.ReplaceAll(strings.ReplaceAll(t, ":", ""), "-", "")
+	filePath = filepath.Join(config.ViamDotDir, fmt.Sprintf("debug_camera_component_%s.txt", t))
+}
+
+// InfoMap is a map of information to be written to the log.
+type InfoMap = map[string]string
+
+// Logger is a thread-safe logger that manages a single log file in config.ViamDotDir.
+type Logger struct {
+	ctx       context.Context
+	nActive   atomic.Int32
+	cancel    func()
+	doneCh    chan struct{}
+	infoCh    chan info
+	Logger    golog.Logger
+	prevInfo  map[string]InfoMap
+	isRunning atomic.Bool
+}
+
+type info struct {
+	title string
+	m     InfoMap
+}
+
+// NewLogger creates a new Logger. No background routines are started until the first entity calls Logger.Add.
+func NewLogger() Logger {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// keeps a rolling log with a size of at most MaxSize
+	lumlog := &lumberjack.Logger{
+		Filename:   filePath,
+		MaxSize:    1, // megabytes
+		MaxBackups: 0, // number of log files
+	}
+
+	lumberjackZapHook := func(e zapcore.Entry) error {
+		_, err := lumlog.Write([]byte(fmt.Sprintf("%+v", e)))
+		utils.UncheckedError(err)
+		return nil
+	}
+
+	cfg := golog.NewDevelopmentLoggerConfig()
+	cfg.OutputPaths = []string{filePath}
+	logger, err := cfg.Build()
+	utils.UncheckedError(err)
+	logger.WithOptions(zap.Hooks(lumberjackZapHook))
+
+	return Logger{
+		ctx:      ctx,
+		cancel:   cancel,
+		doneCh:   make(chan struct{}),
+		infoCh:   make(chan info),
+		prevInfo: make(map[string]InfoMap),
+		Logger:   logger.Sugar().Named("camera_debugger"),
+	}
+}
+
+// Add signals to the logger there's an entity depending on this logger.
+//
+// This function will create the log file if it does not already exist. The filename is unique to the running application.
+// This method is thread safe. See Done.
+func (l *Logger) Add(i int32) {
+	if j := l.nActive.Add(i); j-i != 0 {
+		return
+	}
+
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
+		l.init()
+	}
+
+	utils.PanicCapturingGo(func() {
+		defer close(l.doneCh)
+		defer l.isRunning.Store(false)
+		l.isRunning.Store(true)
+		ticker := time.NewTicker(5 * time.Second)
+		for {
+			select {
+			case <-l.ctx.Done():
+				return
+			case info := <-l.infoCh:
+				l.write(info.title, info.m)
+			case <-ticker.C:
+				l.captureV4L2info()
+			}
+		}
+	})
+}
+
+// Done signals to the logger there's one less entity dependent on this logger.
+//
+// It will block until the logger has shutdown all background threads if and only if it is the last entity watched.
+// This method is thread safe. See Add.
+func (l *Logger) Done() {
+	if j := l.nActive.Add(-1); j == 0 {
+		l.cancel()
```

Comment:
If the last dependency calls Done(), but then a new dependency calls Add(), I believe the context will still be canceled.